osu.Game/Overlays/MusicController.cs

```diff
@@ -243,6 +254,52 @@ private PreviousTrackResult prev()
                 onSuccess?.Invoke();
         });
 
+        /// <summary>
+        /// Attenuates the volume and/or filters the currently playing track.
+        /// </summary>
+        /// <param name="duration">Duration of the ducking transition, in ms.</param>
+        /// <param name="duckVolumeTo">Level to drop volume to (1.0 = 100%).</param>
+        /// <param name="duckCutoffTo">Cutoff frequency to drop `AudioFilter` to. Use `AudioFilter.MAX_LOWPASS_CUTOFF` to skip filter effect.</param>
+        /// <param name="easing">Easing for the ducking transition.</param>
+        public void Duck(int duration = 0, float duckVolumeTo = 0.25f, int duckCutoffTo = 300, Easing easing = Easing.InCubic)
```

Comment:
I do wonder about this API design. While convenient it's also a bit weak on forcing the caller to clean up after themselves (there's nothing stopping anyone from calling `Duck()` and then forgetting to call `Unduck()` later).

The established pattern we sorta have for these is that such transient operations return an `InvokeOnDisposal` that will call `Unduck()` when the `InvokeOnDisposal` is disposed, so rather than do

```csharp
musicController.Duck();
// later
musicController.Unduck();
```

you'd do

```csharp
var duck = musicController.Duck();
// later
duck.Dispose();
```

Although it wold mean that the unduck parameters would need to be set by `Duck()` which may be indesirable.

Also food for thought here is that the current API design allows any number of callers to call `Duck()` and `Unduck()` in arbitrary order concurrently and the last one will win essentially. With the `IDisposable` approach `MusicController` could keep track of who's currently ducking and just not allow anyone else to while that other component is doing it.

Reply:
Yeah I agree this API design doesn't feel great, but nothing better came to mind.

> Also food for thought here is that the current API design allows any number of callers to call `Duck()` and `Unduck()` in arbitrary order concurrently and the last one will win essentially. With the `IDisposable` approach `MusicController` could keep track of who's currently ducking and just not allow anyone else to while that other component is doing it.

This was kinda intentional - ideally triggering `Duck()` and `Unduck()` repeatedly and arbitrarily should be allowed, with the most recent call taking precedence.

In my mind `TimedDuck()` should be preferred for most cases anyway (esp. where usages could overlap), with `Duck()` and `Unduck()` reserved only for when `TimedDuck()` can't be used (i.e. when the `delay` is arbitrary or dependant on the user - like for overlays/popups/etc).

I'll look into `InvokeOnDisposal` though, enforcing `Unduck()` after a `Duck()` is probably worth the clunkiness of passing unduck parameters up front.

Reply:
> This was kinda intentional - ideally triggering `Duck()` and `Unduck()` repeatedly and arbitrarily should be allowed, with the most recent call taking precedence.

If this was intended to be more of an "immediate" mode API then that is fine. For reference I don't consider this review thread a blocker, just vague musings on API shape at most I suppose.

Reply:
It was a valid concern though - I found an edge-case where a `TimedDuck()` is triggered while a `Duck()` is already in progress, resulting in 'premature' unducking, when it should really remain ducked instead.